### PR TITLE
Remove FiniteMathOnly  and use only NoHonorINFs and NoHonorNANs.

### PR DESCRIPTION
**Pull request by zahiraam**
Currently `__FINITE_MATH_ONLY__` is set when `FiniteMathOnly`. And `FiniteMathOnly` is set when `NoHonorInfs` && `NoHonorNans` is true. But what happens one of the latter flags is false? 
To avoid potential inconsistencies, the internal option `FiniteMathOnly` is removed option and the macro `__FINITE_MATH_ONLY__` is set when `NoHonorInfs` && `NoHonorNans`.

### CODE REVIEW

File: clang/include/clang/Basic/LangOptions.h
Comment (AaronBallman) [inline clang/include/clang/Basic/LangOptions.h:823]: How about:
```
assert((LO.FiniteMathOnly == (LO.NoHonorInfs && LO.NoHonorNans)) && "inf/nan inconsistent internal state");
```
to make it more clear that the state has to be sane no matter which options are set? Also, this corrects the duplicate condition in your `if` statement.

Comment (zahiraam) [reply clang/include/clang/Basic/LangOptions.h:823]: With this condition, this command line will assert:
`clang.exe -Xclang -menable-no-infs -Xclang  -menable-no-nans foo.c`
` FiniteMathonly` is `0` by default.
Do we want that?

Comment (AaronBallman) [reply clang/include/clang/Basic/LangOptions.h:823]: That sounds like a logic bug to me -- if infs and nans are disabled, then it's definitely finite math only mode. If infs *OR* nans are disabled... I think it's also finite math only mode.

Ideally, I think we should get rid of the internal notion of `FiniteMathOnly` and replace it with the honor infs/nans flags explicitly so we don't get into these sort of situations to begin with. We'd still have to answer how to define `__FINITE_MATH_ONLY__`, but that seems doable.

CC @andykaylor @jcranmer-intel

Comment (zahiraam) [reply clang/include/clang/Basic/LangOptions.h:823]: There is quite a bit of LIT tests that have that "logic bug" then. They are using` -Xclang -menable-no-infs -Xclang  -menable-no-nan` without the `-ffinite-math-only` option.

Getting rid of `FiniteMathOnly` is fine by me. 
`FiniteMathOnly = !HonorINFs && !HonorNaNs` which would mean `-menable-no-infs -menable-non-nans`. I would think using one or the other of these options could still mean that input values may be either INF or NAN or we can produce them, in which case we wouldn't be in finite math mode.

Comment (AaronBallman) [reply clang/include/clang/Basic/LangOptions.h:823]: > There is quite a bit of LIT tests that have that "logic bug" then. They are using -Xclang -menable-no-infs -Xclang  -menable-no-nan without the -ffinite-math-only option.

I didn't mean the logic bug was on the user end; I meant that our internal state has the logic bug because `!FiniteMathOnly` implies that infinities are supported and so it logically conflicts with `!HonorInfs`.

> Getting rid of FiniteMathOnly is fine by me.
FiniteMathOnly = !HonorINFs && !HonorNaNs which would mean -menable-no-infs -menable-non-nans. I would think using one or the other of these options could still mean that input values may be either INF or NAN or we can produce them, in which case we wouldn't be in finite math mode.

I can see the logic there and I think that's defensible, but because `FiniteMathOnly` determines the value `__FINITE_MATH_ONLY__` expands to, we should probably see how that flag is being used in the wild to see whether folks expect it to be `1` or `0` when infinity is enabled but NAN is disabled (for example).

Comment (andykaylor) [reply clang/include/clang/Basic/LangOptions.h:823]: I agree with Aaron that we should get rid of the internal state that combines these two modes. I expect we'll only want to define `__FINITE_MATH_ONLY__` when we aren't honoring either. You could create code that made incorrect assumptions either way, but the state described by the symbol is that we aren't honoring either.

Comment (jcranmer-intel) [reply clang/include/clang/Basic/LangOptions.h:823]: The way I see it, nnan and ninf collectively define four modes for math: full types, no-nan, finite-math (nnan + ninf), and the extremely questionable (to me at least) ninf-without-nnan. I don't think `FiniteMathOnly` as a concept independent of the nnan/ninf flags makes much sense.

There is perhaps room for disagreement as to whether `__FINITE_MATH_ONLY__` should be `nnan || ninf` or `nnan && ninf`; I lean towards && myself. Looking at sourcegraph, there seems to be essentially three uses of these macros: two to indicate that they don't care about min(NaN) results, and one to avoid using fpclassify.

File: clang/include/clang/Basic/LangOptions.h
Comment (AaronBallman) [inline clang/include/clang/Basic/LangOptions.h:822]: The `FiniteMathOnly` variable should be removed from `LangOptions` entirely at this point; the plan is to only rely on `NoHonorInfs` and `NoHonorNans` instead.

Comment (zahiraam) [reply clang/include/clang/Basic/LangOptions.h:822]: There is no need for the assertion then. I suppose `CLFiniteMathOnly` should be removed too, right?

Comment (AaronBallman) [reply clang/include/clang/Basic/LangOptions.h:822]: Correct, that assertion can go away (don't forget to update the PR title and summary accordingly as well).

>  I suppose CLFiniteMathOnly should be removed too, right?

CC @AnastasiaStulova as OpenCL code owner.

I think `CLFiniteMathOnly` is a combination of `NoHonorInfs`, `NoHonorNaNs`, and `OpenCL` language mode. However, I don't see `CLFiniteMathOnly` being *used* by anything when I grep over the source. We set the flag from Options.td via marshaling, but we don't seem to ever check the value of the LangOpt anywhere. So it seems like `CLFiniteMathOnly` can be removed and maybe `-cl-finite-math-only` could be deprecated and eventually removed?

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:3298]: `InfValues` isn't initialized here

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Clang.cpp:3307]:
```suggestion
  for (auto *Arg : Args.filtered(options::OPT_Xclang))
    processArg(Arg);
```

File: clang/lib/Basic/LangOptions.cpp
Comment (AaronBallman) [inline clang/lib/Basic/LangOptions.cpp:38]: I think we don't need this code at all because `LangOptions.def` defines both of these with a default value of 0 anyway, so the x macros above should already do this automatically, right?

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (AaronBallman) [inline clang/lib/Driver/ToolChains/Clang.cpp:3310]: Instead of filtering all arguments to cc1 and looping over all of them, can we use `Args.getLastArg()`? And can we skip that processing if we already know we're doing to pass `-ffinite-math-only` because of `(!HonorINFs && !HonorNaNs)`?

Comment (zahiraam) [reply clang/lib/Driver/ToolChains/Clang.cpp:3310]: I can't use `Args.getLastArg()` with arguments to `Xclang`. `Args.filtered()` needs to be used.

Comment (AaronBallman) [reply clang/lib/Driver/ToolChains/Clang.cpp:3310]: Okay, thanks! We should still skip processing those args when we can though.

Comment (AaronBallman) [reply clang/lib/Driver/ToolChains/Clang.cpp:3310]: That could work, I was thinking along these lines:
```
  // Explanation of why we need to do this dance goes here.
  auto NaNsAndInfs = [&] {
    bool InfValues = true;
    bool NanValues = true;
    auto processArg = [&](const auto *Arg) {
      if (StringRef(Arg->getValue()) == "-menable-no-nans")
        NanValues = false;
      if (StringRef(Arg->getValue()) == "-menable-no-infs")
        InfValues = false;
    };

    for (auto *Arg : Args.filtered(options::OPT_Xclang))
      processArg(Arg);

    return InfValues && NanValues;
  };

  if ((!HonorINFs && !HonorNaNs) || !NaNsAndInfs())
    CmdArgs.push_back("-ffinite-math-only");
```
I leave it to driver folks to say what they'd prefer if they have a strong preference.

Comment (zahiraam) [reply clang/lib/Driver/ToolChains/Clang.cpp]: @mdtoguchi WDYT?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp]: No strong preference - singular addition of `-ffinite-math-only` is easier to maintain.

Comment (zahiraam) [reply clang/lib/Driver/ToolChains/Clang.cpp]: I see. I could write something like this as a compromise?
`bool shouldAddFiniteMathOnly = false;`

`if (!HonorINFs && !HonorNaNs) {`
  `shouldAddFiniteMathOnly = true;`
`} else {`
  `bool InfValues = true;`
 `bool NanValues = true;`

  `for (const auto *Arg : Args.filtered(options::OPT_Xclang)) {`
    `StringRef ArgValue = Arg->getValue();`
    `if (ArgValue == "-menable-no-nans")`
      `NanValues = false;`
    `else if (ArgValue == "-menable-no-infs")`
      `InfValues = false;`
  `}`

 ` if (!NanValues && !InfValues)`
   `shouldAddFiniteMathOnly = true;`
`}`

`if (shouldAddFiniteMathOnly) {`
  `CmdArgs.push_back("-ffinite-math-only");`
`}`

Comment (AaronBallman) [reply clang/lib/Driver/ToolChains/Clang.cpp]: I'd be okay with that approach.

File: clang/lib/Basic/Targets/OSTargets.cpp
Comment (jcranmer-intel) [inline clang/lib/Basic/Targets/OSTargets.cpp:177]: Given that NoHonorInfs and NoHonorNaNs are both individually checked below, you don't need this entry here.

**Comment by AaronBallman** [comment]
Ping @AnastasiaStulova for OpenCL feedback (this PR impacts an OpenCL command line option).

**Review by AaronBallman — APPROVED** [review]
Changes LGTM, I've added @MaskRay and @jansvoboda11 in case they can help answer the OpenCL questions, as it seems @AnastasiaStulova may be unavailable at the moment.

If we don't hear back from anyone in the next two days, I think it's fine to land.

File: clang/include/clang/Basic/LangOptions.h
Comment (AnastasiaStulova) [reply clang/include/clang/Basic/LangOptions.h]: Btw this option is defined in OpenCL spec so it would be better to keep this at least as a driver option.
https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#optimization-options

I think mapping it to ``NoHonorInfs``, ``NoHonorNaNs`` frontend options seems reasonable.

Can the description of this change be elaborated to describe what the intension is.

Tagging @svenvh for any additional feedback.

 But mapping

Comment (zahiraam) [reply clang/include/clang/Basic/LangOptions.h]: @AnastasiaStulova thanks for the review.
The driver option `-cl-finite-math-only` can still be used.
Updated the PR's description.

File: clang/test/CodeGenOpenCL/relaxed-fpmath.cl
Comment (zahiraam) [reply clang/test/CodeGenOpenCL/relaxed-fpmath.cl]: Done.

**Review by svenvh — APPROVED** [review]

**Pull request merged**
